### ISSUE TITLE
Audio interpolator 

### ISSUE THREAD

**Issue by Theoldnation** [Arisotura/melonDS#901]
Maybe should be good to have an audio interpolator like desmume has
without and audio interpolator some games sometimes sound weird and really low quality in comparation of desmume and his
audio interpolator

thanks for your attention

**Comment by RSDuck** [comment]
that should only matter if the games are running sub fullspeed where the experience is meh anyway. Have you tried enabling the JIT?

**Comment by RSDuck** [comment]
oh you're talking about the upsampling to get from the native sample rate of the DS to what ever will be output?

**Comment by Theoldnation** [comment]
> oh you're talking about the upsampling to get from the native sample rate of the DS to what ever will be output?

Yeah, because my issue isn't related to games no running at full speed, even on full speed the games sound very low quality in comparison of desmume

**Comment by Arisotura** [comment]
they mean applying linear/whatever interpolation to output from the various DS channels

which might doable by using the channel timers, but require a bit of cleverness to work properly

**Comment by nadiaholmquist** [comment]
This is more of a feature request than an issue, as if you listen to games on a real DS you'll notice that DS games just have pretty crusty audio.

Still a good idea though.

**Comment by RSDuck** [comment]
I think this issue breaks down to two parts. The first one is for interpolation inside the SPU while playing back samples. And the other one is for the conversion from the internal sampling rate of the SPU to usually 48khz what the audio APIs require. We currently use nearest filtering for the latter. I'm not that familiar with signal processing stuff, but I think we're definitely loosing some quality in that step, which would still be there if the signal of the SPU was straight passed to a DAC like on a real DS.

**Comment by Braok** [comment]
> I think this issue breaks down into two parts. The first one is for interpolation inside the SPU while playing back samples. And the other one is for the conversion from the internal sampling rate of the SPU to usually 48khz what the audio APIs require. We currently use nearest filtering for the latter. I'm not that familiar with signal processing stuff, but I think we're definitely loosing some quality in that step, which would still be there if the signal of the SPU was straight passed to a DAC like on a real DS.

I was about to make a thread on this because it's definitely bothering me. But yes, I think that the easiest, and most accurate solution, would be to use nearest up to 32728hz like on a real DS and then resample to the device sample rate using a higher quality resampler (linear, sinc, whatever). Audio quality would boost tremendously. It definitely weirded me out when I popped in a game with 32728hz music that sounds great on a real DS but makes my ears bleed on MelonDS. I think this would be a better approach than desmume anyhow (there are filter options, but i think it affects all sounds, so unless a game exclusively uses 32khz samples then it'll all end up filtered even though it shouldn't be)

anyhow it shouldn't be a crazy change, and the option to switch filter modes is always available for low-end systems

**Comment by poudink** [comment]
32728hz? wasn't it 32768?

**Comment by Braok** [comment]
> 32728hz? wasn't it 32768?

The DS and 3DS are odd and use 32728hz instead of 32768hz. This is why so many games use 16364hz samples.

**Comment by BestestHD** [comment]
I wanna update this by confirming that there's definitely something else going on. I think MelonDS is going at the correct sample rate, so the audio filter is the issue. Popped in a game modded to use a 48khz stream which still sounds crunchy despite my sound card being set to 48khz.

Would it not be better to even use the operating system's built-in interpolation? I understand that using a different audio backend like OpenAL is a bit much, but a better solution would go a long way. DS games sound crunchy, but only to a certain extent.
I'm going to try figuring out something, though I have no clue how to work with C++ 💀 Worth a shot, though.

**Comment by RSDuck** [comment]
> I wanna update this by confirming that there's definitely something else going on. I think MelonDS is going at the correct sample rate, so the audio filter is the issue. Popped in a game modded to use a 48khz stream which still sounds crunchy despite my sound card being set to 48khz.

in the end the DS produces a stream at a 32khz, you're 48khz signal probably is going to be downsampled.

> Would it not be better to even use the operating system's built-in interpolation? 

because that can lead to some odd problems. With SDL (what we use to present audio) it's possible to ask whether a certain sample rate is supported. So 1. the odd DS samplerate might not even be supported and 2. (that's why we stopped doing that iirc) sometimes they would lie and say that samplerate is supported and then not output any sound at all.

**Comment by Arisotura** [comment]
it's even worse than that, assuming it runs at fullspeed consistently, melonDS outputs at a non-integer sample rate (560190 cycles per frame * 60 frames per second / 1024 cycles per sample = 32823.6328125 Hz). there's no way you can get that exact sample rate. early melonDS picked the closest integer rate and prayed, which didn't always work well. that's why it has its built-in resampler now (plus, said resampler has some leeway to make up for small speed fluctuations).

**Comment by BestestHD** [comment]
> in the end the DS produces a stream at a 32khz, you're 48khz signal probably is going to be downsampled.
This was in reference to the emulators DraStic and DeSmuME, which use the system rate. The same 48khz sample sounds great in them as long as the PC/phone is set to 48khz.
> > Would it not be better to even use the operating system's built-in interpolation?
> 
> because that can lead to some odd problems. With SDL (what we use to present audio) it's possible to ask whether a certain sample rate is supported. So 1. the odd DS samplerate might not even be supported and 2. (that's why we stopped doing that iirc) sometimes they would lie and say that samplerate is supported and then not output any sound at all.

I see, that makes sense. 32728hz has probably only ever been used by the 3DS and DS😅Speaking of 3DS, Citra uses an internal sampling rate of 32728hz and sounds fine. There's a fine-sounding SDL2 implementation as well, [so maybe Citra's code could be used as a reference here?](https://github.com/citra-emu/citra/blob/master/src/audio_core/interpolate.cpp)

> melonDS outputs at a non-integer sample rate (560190 cycles per frame * 60 frames per second / 1024 cycles per sample = 32823.6328125 Hz)

Also noticed this number while messing with melonDS (to no avail). Not gonna begin to question why it's outputting above the native rate, and such an odd number at that, but it is what it is I guess. Timing stuff is all way beyond my understanding 💀

**Comment by Arisotura** [comment]
the native rate is prolly dependent on the system clock, as that is what drives the audio subsystem.

**Comment by Arisotura** [comment]
also: the DS runs slightly below 60FPS (59.82 or something). as everything is tied together timing-wise, that makes the audio output rate closer to 32728Hz.

melonDS outputs at a higher rate because it tries to run at 60FPS, which is slightly faster than the real thing. in reality, the output rate depends on how fast the whole thing runs (just like how altering the DS's system clock changes the audio pitch, as shown here: https://www.youtube.com/watch?v=PYIvASbELkw ).
